packages/core/src/render3/util/discovery_utils.ts

```diff
@@ -203,6 +218,61 @@ export function getDirectives(element: Element): {}[] {
   return context.directives === null ? [] : [...context.directives];
 }
 
+/**
+ * Metadata for a directive developers can use for debugging purposes or tooling.
+ */
+export interface DirectiveMetadata {
+  inputs: Record<string, string>;
+  outputs: Record<string, string>;
+}
+
+/**
+ * Metadata for a component developers can use for debugging purposes or tooling.
+ */
+export interface ComponentMetadata extends DirectiveMetadata {
+  encapsulation: ViewEncapsulation;
+  onPush: boolean;
+}
+
+/**
+ * Returns the metadata for a particular directive or component. The function accepts an instance
+ * of a directive or component and returns the corresponding metadata.
+ *
+ * @param directiveOrComponentInstance Instance of a directive or component
+ * @returns metadata of the passed directive or component
+ */
+export function getDirectiveMetadata(directiveOrComponentInstance: any): ComponentMetadata|
+    DirectiveMetadata|null {
+  const {constructor} = directiveOrComponentInstance;
+  if (!constructor) {
+    throw new Error('Unable to find the instance constructor');
+  }
+  const directiveDef = getDirectiveDef(constructor);
+  if (directiveDef) {
+    return {inputs: directiveDef.inputs, outputs: directiveDef.outputs};
+  }
+  const componentDef = getComponentDef(constructor);
+  if (componentDef) {
+    return {
+      inputs: componentDef.inputs,
+      outputs: componentDef.outputs,
+      encapsulation: componentDef.encapsulation,
+      onPush: componentDef.onPush
+    };
+  }
+  return null;
+}
+
+/**
+ * Returns a `DebugNode` for a specified DOM node.
+ *
+ * @param node a DOM node the method will wrap in a DebugNode
+ * @returns a debug node
+ */
+export function asDebugNode(node: Node): PublicDebugNode {
+  return getDebugNode__POST_R3__(node);
+}
```

Comment:
Should we create an instance of `DebugElement` (that has more methods) instead of `DebugNode`?
I'd also propose to rename the function from `asDebugNode` to `createDebugNode` - that might be a bit easier to read/understand.
```suggestion
/**
 * Creates an instance of `DebugNode` class for a specified DOM node.
 *
 * @see `DebugNode`
 *
 * @param node a DOM node that will be wrapped into a `DebugNode` instance
 * @returns an `DebugNode` instance
 */
export function createDebugNode(node: Node): DebugNode {
  return getDebugNode__POST_R3__(node);
}
```